returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,251 @@
+    speed perturbation into the data loading pipeline.
+
+    The integration into the data loading pipeline makes it easy to distribute the
+    data processing work across multiple CPU cores using `MultiProcDataset` and in
+    turn frees the GPU from data preprocessing tasks.
+
+    Both functions can be specified at the same time, and their results will be
+    combined by first applying the segment-level function and then the
+    multiple-segments-level function.
+
+    Example usage:
+        train = {
+            "class": "PostprocessingDataset",
+            "dataset": {
+                "class": "HDFDataset",
+                "files": ["/path/to/data.hdf"],
+            },
+            # at least one of them:
+            "map_seq": map_seq,  # (data: TensorDict) -> TensorDict
+            "map_seq_stream": map_seqs,  # (iter: Iterator[TensorDict]) -> Iterator[TensorDict]
+        }
+    """
+
+    def __init__(
+        self,
+        dataset: Dict[str, Any],
+        map_seq: Optional[
+            Union[
+                Callable[[TensorDict], TensorDict],
+                Callable[[TensorDict], Iterator[TensorDict]],
+            ]
+        ] = None,
+        map_seq_stream: Optional[Callable[[Iterator[TensorDict]], Iterator[TensorDict]]] = None,
+        map_outputs: Optional[Dict[str, Any]] = None,
+        _meta_info_cache: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+        """
+        :param dataset: inner dataset to be post-processed
+        :param map_seq: post processor function operating on the single-segment level.
+        :param map_seq_stream: post processor function operating on the multiple segment level via an iterator.
+            Allows merging multiple segments into one, or generating multiple output segments from one input segment.
+            When both functions are specified, the segment-level function is applied first to every segment, and the
+            results are passed to the iterator-level function.
+        :param output: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+        :param _meta_info_cache: internal usage
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        assert (
+            self.seq_ordering == "default"
+        ), f"specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}"
+        assert (
+            map_seq is not None or map_seq_stream is not None
+        ), "need to either define map_seq or map_seq_stream functions"
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq or self._identity
+        self._map_seq_stream = map_seq_stream or self._identity
```

Comment:
```suggestion
        self._map_seq = map_seq
        self._map_seq_stream = map_seq_stream
```
See also my comments above on allowing both (I would not allow both), and comments below on `_wrap_map_seq`.